### PR TITLE
Error on types named "undefined"

### LINKED ISSUE #57573: A type can be named undefined
### 🔎 Search Terms

type named undefined

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ 

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEwpIN4FgBQS9IgA8YB7AJwBckKBPGZAV2ABMIAzAS2AmaQF4kAIkEBuHPgLFyVNk3gUOJYEkIAKGAC4kTVp27MAlFp3suPDEgD0lpAHMANiQBGIe9pan94ifm94AvjiB2DiywPKKymqaKFAAdCZ6PAYW1u66ZrxsZCRg1AAWyKCQsAgQOEEV2EA)

### 💻 Code

```ts
namespace ns {
    export type undefined = "";
    export function x(p: undefined): undefined { // global undefined
        
    }
}

function x(p: ns.undefined) { // undefined from the namespace

}
```

### 🙁 Actual behavior

Inside the namespace `undefined` is the global `undefined` type. We can access the type if we qualify it with the namespace

### 🙂 Expected behavior

`undefined` should not be allowed as a type name. (`null` is not)

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #57573

This makes us error on types named `undefined` in the same way we do keyword types. Though undefined is not a keyword, it _is_ a keyword in type space.

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test top400
@typescript-bot user test this
@typescript-bot run dt

**Comment by jakebailey** [comment]
FWIW the way I wanted to implement this was "if you see an `undefined` Identifier in type space (besides a typeof), error", but couldn't figure out the best way to do that, so I just threw errors onto everything that introduced a name to type space instead...

**Comment by jakebailey** [comment]
Good, nobody!

**Comment by jakebailey** [comment]
Well, given #57573 is now effectively accepted, I guess I'll clean this up?

**Comment by jakebailey** [comment]
Apparently I'm dumb and didn't realize that we already issue this error:

```
==== undefinedTypeAssignment4.ts (6 errors) ====
    class undefined {
         ~~~~~~~~~~
!!! error TS2397: Declaration name conflicts with built-in global identifier 'undefined'.
          ~~~~~~~~~
!!! error TS2397: Declaration name conflicts with built-in global identifier 'undefined'.
    	foo: string;
    }
    interface undefined {
```

I just added a duplicate one. The actual problem are the other ones we're missing.

**Comment by jakebailey** [comment]
Yeah, I did not test what I thought I was.

**Comment by jakebailey** [comment]
Okay, I _sorta_ tested the right thing. This check is duplicative, but I didn't ignore this in globals, so I did actually check the right thing, kinda.

**Comment by jakebailey** [comment]
@typescript-bot test top400
@typescript-bot user test this
@typescript-bot run dt

**Comment by jakebailey** [comment]
@typescript-bot test top400
@typescript-bot user test this
@typescript-bot run dt

File: tests/baselines/reference/undefinedTypeAssignment2.errors.txt [hunk 1]
```diff
@@ -1,8 +1,8 @@
-undefinedTypeAssignment2.ts(1,5): error TS2397: Declaration name conflicts with built-in global identifier 'undefined'.
+undefinedTypeAssignment2.ts(1,5): error TS2300: Duplicate identifier 'undefined'.
 
 
 ==== undefinedTypeAssignment2.ts (1 errors) ====
     var undefined = void 0;
         ~~~~~~~~~
-!!! error TS2397: Declaration name conflicts with built-in global identifier 'undefined'.
+!!! error TS2300: Duplicate identifier 'undefined'.
```
Comment (jakebailey) [inline tests/baselines/reference/undefinedTypeAssignment2.errors.txt:7]: This is probably not a great error now that I dropped the builtinGlobals method of doing this. Probably I should keep that code around but only error in value space.

File: tests/baselines/reference/undefinedTypeAssignment3.errors.txt [hunk 2]
```diff
@@ -1,8 +0,0 @@
-undefinedTypeAssignment3.ts(1,5): error TS2397: Declaration name conflicts with built-in global identifier 'undefined'.
-
-
-==== undefinedTypeAssignment3.ts (1 errors) ====
-    var undefined = null;
```
Comment (jakebailey) [inline tests/baselines/reference/undefinedTypeAssignment3.errors.txt:5]: I'm brainfarting and failing to find the right function to still emit this error on non-types. Will figure out later when my brain is working...

**Comment by jakebailey** [comment]
@typescript-bot test top400
@typescript-bot user test this
@typescript-bot run dt

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by jakebailey** [comment]
Okay, this should be ready. Removed the namespace stuff as discussed.

**Comment by jakebailey** [comment]
@typescript-bot test top400
@typescript-bot user test this
@typescript-bot run dt

@typescript-bot perf test this faster
@typescript-bot pack this

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2720,20 +2717,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
     }
 
-    function addToSymbolTable(target: SymbolTable, source: SymbolTable, message: DiagnosticMessage) {
```
Comment (jakebailey) [inline src/compiler/checker.ts:2723]: I opted to eliminate builtinGlobals while here; its only use is to complain about undefined, but now we're going to complain even less and the new code had to exclude type declarations anyway.

**Comment by jakebailey** [comment]
Abusing this PR to try out a cherry-pick with the new bot.

@typescript-bot cherry-pick this to release-5.4

**Review by weswigham — APPROVED** [review]

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -2720,20 +2717,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
-        });
-
-        function addDeclarationDiagnostic(id: string, message: DiagnosticMessage) {
-            return (declaration: Declaration) => diagnostics.add(createDiagnosticForNode(declaration, message, id));
+    function addUndefinedToGlobalsOrErrorOnRedeclaration() {
+        const name = undefinedSymbol.escapedName;
+        const targetSymbol = globals.get(name);
+        if (targetSymbol) {
+            forEach(targetSymbol.declarations, declaration => {
+                // checkTypeNameIsReserved will have added better diagnostics for undefined.
```
Comment (jakebailey) [inline src/compiler/checker.ts:2725]:
```suggestion
                // checkTypeNameIsReserved will have added better diagnostics for type declarations.

```

**Pull request merged**